src/polymer/element-before-dom-module.ts

```diff
@@ -0,0 +1,133 @@
+
+/**
+ * Unbalanced binding expression delimiters occurs when a value such as
+ * `[[myValue]]` or `{{myValue}}` have too many or too few brackets on either
+ * side.
+ */
+export class ElementBeforeDomModule extends HtmlRule {
+  readonly code = 'element-before-dom-module';
+  readonly description = stripIndent(`
+      Warns for an element being declared before its dom-module.
+
+      For example, this is invalid:
+        <script>Polymer({is: 'my-elem'})</script>
+        <dom-module id='my-elem'></dom-module>
+
+      But this is fine:
+        <dom-module id='my-elem'></dom-module>
+        <script>Polymer({is: 'my-elem'})</script>
+      `);
+
+  constructor() {
+    super();
+  }
+
+  public async checkDocument(
+      parsedHtml: ParsedHtmlDocument, document: Document): Promise<Warning[]> {
+    const warnings: Warning[] = [];
+
+    const domModules = document.getByKind('dom-module');
+    if (domModules.size === 0) {
+      return warnings;  // Early exit quick in the trivial case.
+    }
+    const domModulesByTagName = new Map<string, DomModule>();
+    for (const domModule of domModules) {
+      if (!domModule.id) {
+        continue;
+      }
+      domModulesByTagName.set(domModule.id, domModule);
+    }
+
+    // Gather together all elements defined in this file and in direct imports.
+    // Group them together with the source range in this file that they first
+    // become active. (We don't look into transitive imports because circularity
+    // makes ordering complicated.)
+    const localElements =
+        Array.from(document.getByKind('polymer-element'))
+            .filter((el) => !!el.sourceRange)
+            .map((el) => ({sourceRange: el.sourceRange!, elements: [el]}));
+    const elementsByImport =
+        Array.from(document.getByKind('import'))
+            .filter((i) => i.sourceRange)
+            .map((i) => {
+              if (!i.document) {
+                return undefined!;
+              }
+              // For complicated reasons, non-module script src tags are kinda
+              // treated like independent documents, and kinda like inline
+              // scripts. Long story short, we need to make sure that any
+              // elements defined "in them" aren't actually defined in us, their
+              // importer.
+              const elements =
+                  Array.from(i.document.getByKind('polymer-element'));
+              const nonlocalElements = elements.filter(
+                  (e) => e.sourceRange && e.sourceRange.file !== document.url);
+              return {sourceRange: i.sourceRange!, elements: nonlocalElements};
+            })
+            .filter((v) => !!v);
+
+    // Sort the element groups by the order in which they appear in the
+    // document.
+    const sorted = localElements.concat(elementsByImport).sort((a, b) => {
+      return comparePositionAndRange(a.sourceRange.start, b.sourceRange);
+    });
+
+    const seenSoFar = new Set<null|string>([null]);
```

Comment:
Good catch, from an earlier version of the code where that's how I was dealing with unknown tag names. Fixed.